src/libp2p_crypto.erl

```diff
@@ -2,119 +2,155 @@
+         bin_to_b58/1, bin_to_b58/2,
+         b58_to_bin/1, b58_to_version_bin/1,
+         pubkey_to_b58/1, b58_to_pubkey/1,
+         pubkey_bin_to_p2p/1, p2p_to_pubkey_bin/1,
+         verify/3
         ]).
 
--spec make_public_key(private_key()) -> public_key().
-make_public_key(#'ECPrivateKey'{parameters=Params, publicKey=PubKey}) ->
-    {#'ECPoint'{point=PubKey}, Params}.
-
 %% @doc Generate keys suitable for a swarm.  The returned private and
 %% public key has the attribute that the public key is a compressable
 %% public key.
--spec generate_keys() -> {private_key(), public_key()}.
-generate_keys() ->
+-spec generate_keys(key_type()) -> key_map().
+generate_keys(ecc_compact) ->
     {ok, PrivKey, CompactKey} = ecc_compact:generate_key(),
     PubKey = ecc_compact:recover_key(CompactKey),
-    {PrivKey, PubKey}.
+    #{secret => {ecc_compact, PrivKey}, public => {ecc_compact, PubKey}};
+generate_keys(ed25519) ->
+    #{public := PubKey, secret := PrivKey} = enacl:crypto_sign_ed25519_keypair(),
+    #{secret => {ed25519, PrivKey}, public => {ed25519, PubKey}}.
+
+
 
 %% @doc Load the private key from a pem encoded given filename.
 %% Returns the private and extracted public key stored in the file or
 %% an error if any occorred.
--spec load_keys(string()) -> {ok, private_key(), public_key()} | {error, term()}.
+-spec load_keys(string()) -> {ok, key_map()} | {error, term()}.
 load_keys(FileName) ->
     case file:read_file(FileName) of
-        {ok, PemBin} -> from_pem(PemBin);
+        {ok, Bin} -> {ok, keys_from_bin(Bin)};
         {error, Error} -> {error, Error}
     end.
 
--spec mk_sig_fun(private_key()) -> sig_fun().
-mk_sig_fun(PrivKey) ->
-    fun(Bin) -> public_key:sign(Bin, sha256, PrivKey) end.
+-spec mk_sig_fun(privkey()) -> sig_fun().
+mk_sig_fun({ecc_compact, PrivKey}) ->
+    fun(Bin) -> public_key:sign(Bin, sha256, PrivKey) end;
+mk_sig_fun({ed25519, PrivKey}) ->
+    fun(Bin) -> enacl:sign_detached(Bin, PrivKey) end.
+
 
 %% @doc Store the given keys in a file.  See @see key_folder/1 for a
 %% utility function that returns a name and location for the keys that
 %% are relative to the swarm data folder.
--spec save_keys({private_key(), public_key()}, string()) -> ok | {error, term()}.
-save_keys({PrivKey, _PubKey}, FileName) when is_list(FileName) ->
-    PemBin = to_pem(PrivKey),
-    file:write_file(FileName, PemBin).
+-spec save_keys(key_map(), string()) -> ok | {error, term()}.
+save_keys(KeysMap, FileName) when is_list(FileName) ->
+    Bin = keys_to_bin(KeysMap),
+    file:write_file(FileName, Bin).
 
--spec to_pem(private_key()) -> binary().
-to_pem(PrivKey) ->
+-spec keys_to_bin(key_map()) -> binary().
+keys_to_bin(#{secret := {ecc_compact, PrivKey}, public := {ecc_compact, _PubKey}}) ->
     PemEntry = public_key:pem_entry_encode('ECPrivateKey', PrivKey),
-    public_key:pem_encode([PemEntry]).
+    <<?KEYTYPE_ECC_COMPACT:8, (public_key:pem_encode([PemEntry]))/binary>>;
```

Comment:
You're mixing binary and ascii data here, this seems inadviseable?

Reply:
Maybe it's better to just do a binary format similar to the 25519 curve?

Reply:
but isn't `pem_encode(PemEntries) -> binary()` the right type signature for creating a binary here?

Reply:
Resolved with a87effe